Consider Strictness Of Dependencies in Graph Traversal

cowlingjosh opened the pull request:
Currently we only consider the strictness of dependencies when building a TaskGraph, not during the execution of it. 

Consider two groups A and B both of which are made of two tasks, where B is dependent on A but non-strictly. When running `kraken run A B`, we will build a TaskGraph which looks like:

```mermaid
flowchart TD
    A(Task 1) -->B[A]
    C(Task 2) -->B[A]

    B -.-> D
    B -.-> F
    B -.-> E
    D(Task 3) --> E[B]
    F(Task 3) --> E[B]

```
where the dotted lines and solid lines represent non-strict and strict dependencies respectively. With the current implementation Tasks 1 and 2 will be queued first but if any of them fail, then we will not execute Tasks 3 or 4 due to them not having a in-degree of 0 as we will not remove the GroupTask node A. This is sub-optimal as the non-strict dependency should "only dictates the order of tasks if both were to be executed (and prevents the task from being executed in parallel)." and not block execution.

This PR adds the logic of removing any non-strict dependency (edge) when a task fails, to allow all possible tasks to execute during a `kraken run` whilst still returning an error on the failed tasks.

NiklasRosenstein commented on src/kraken/core/system/graph.py:
```diff
@@ -158,8 +161,34 @@ def _remove_nodes_keep_transitive_edges(self, nodes: Iterable[str]) -> None:
+        removable_edges: set[tuple[str, str]] = set()
+
+        def set_non_strict_edge_for_removal(u: str, v: str) -> None:
+            nonlocal removable_edges
```
this would only be needed if we'd actually want to _set_ the `removal_edges` variable, access to it does not require the `nonlocal` statement.

NiklasRosenstein commented on src/kraken/core/system/graph.py:
```diff
@@ -158,8 +161,34 @@ def _remove_nodes_keep_transitive_edges(self, nodes: Iterable[str]) -> None:
+
+        for failed_task_path in self._failed_tasks:
+            for out_task_path in self._digraph.successors(failed_task_path):
+                out_task_data = self._digraph.nodes[out_task_path]["data"]
```
```suggestion
                out_task = self._digraph.nodes[out_task_path]["data"]
```

nit: Would suggest to just call this `out_task` since that resembles the type of the variable more concretely.

NiklasRosenstein commented on tests/test_task_graph.py:
```diff
@@ -218,3 +218,79 @@ def test__TaskGraph__test_inverse_group_relationship(kraken_project: Project, in
+
+def test__TaskGraph__allow_subsequent_group_execution_on_non_strict_failed_tasks(kraken_project: Project) -> None:
+    """Tests that the TaskGraph correctly allows the graph execution to continue if non-strict dependencies fail, but
+    correctly raises that the graph is not complete.
```
nit: "raises" is not technically the correct term because no exception is raised, just a boolean flag is returned by `graph.is_complete()`

cowlingjosh replied:
Valid nit, have changed to "returns"

NiklasRosenstein reviewed (commented):
This is a great PR, thanks @cowlingjosh! Great test coverage as well. I've left a few notes, but overall this seems good to go!

cowlingjosh commented:
> This is a great PR, thanks @cowlingjosh! Great test coverage as well. I've left a few notes, but overall this seems good to go!

Thanks @NiklasRosenstein! Have just addressed your comments 😃
